include/circt/Scheduling/Scheduler.h

```diff
@@ -0,0 +1,253 @@
+protected:
+  bool hasOperation(Operation *op) { return ops.contains(op); }
+  const OpSet &getOperations() { return ops; }
+
+  bool hasDependence(Dependence *dep) { return deps.contains(dep); }
+  const DepSet &getDependences() { return deps; }
+
+  bool hasOperatorType(OperatorType *opr) { return oprs.contains(opr); }
+  const OprSet &getOperatorTypes() { return oprs; }
+
+  //===--------------------------------------------------------------------===//
+  // Subclass access to properties: retrieve problem, set solution
+  //===--------------------------------------------------------------------===//
+protected:
+  bool hasAssociatedOperatorType(Operation *op) { return assocOpr.count(op); }
+  OperatorType *getAssociatedOperatorType(Operation *op) {
+    return assocOpr.lookup(op);
+  }
+
+  bool hasLatency(OperatorType *opr) { return latency.count(opr); }
+  unsigned getLatency(OperatorType *opr) { return latency.lookup(opr); }
+
+  void setStartTime(Operation *op, unsigned val) { startTime[op] = val; }
+
+  //===--------------------------------------------------------------------===//
+  // Client access to properties: specify problem, retrieve solution
+  //===--------------------------------------------------------------------===//
+public:
+  void setAssociatedOperatorType(Operation *op, OperatorType *opr) {
+    assocOpr[op] = opr;
+  }
+
+  void setLatency(OperatorType *opr, unsigned val) { latency[opr] = val; }
+
+  bool hasStartTime(Operation *op) { return startTime.count(op); }
+  unsigned getStartTime(Operation *op) { return startTime.lookup(op); }
+
+  //===--------------------------------------------------------------------===//
+  // Hooks to check/verify the different problem components
+  //===--------------------------------------------------------------------===//
+protected:
+  virtual LogicalResult checkOp(Operation *op);
+  virtual LogicalResult checkDep(Dependence *dep);
+  virtual LogicalResult checkOpr(OperatorType *opr);
+  virtual LogicalResult checkProb();
+
+  virtual LogicalResult verifyOp(Operation *op);
+  virtual LogicalResult verifyDep(Dependence *dep);
+  virtual LogicalResult verifyOpr(OperatorType *opr);
+  virtual LogicalResult verifyProb();
+
+public:
+  /// Return success if the constructed scheduling problem is valid.
+  LogicalResult check();
+  /// Return success if the computed solution is valid.
+  LogicalResult verify();
+
+  //===--------------------------------------------------------------------===//
+  // Entry point to actual algorithm
+  //===--------------------------------------------------------------------===//
+public:
+  // Attempt to schedule the constructed problem.
+  virtual LogicalResult schedule() = 0;
+};
+
+/// This class models a *dependence* from the `srcIdx`'th result of `src` to the
+/// `destIdx`'th operand of `dst`. In other words, it represents an edge in a
+/// scheduling problem.
+class Dependence : public mlir::StorageUniquer::BaseStorage {
```

Comment:
I think this can be modeled by a tuple of `OpResult`, `OpOperand`.

Reply:
Also, how can this be used to model out-of-SSA dependencies? (e.g. memory loads/stores) Or are you not worrying about that yet?

Reply:
> Or are you not worrying about that yet?

No, quite the opposite, I'd really like to have a uniform modeling for the edges of the SSA graph as well as for memory dependencies. I played around a bit with `OpResult`/`OpOperand`, but these seem tightly knit into the SSA graph, and I currently don't see how that could work for memory dependences.

Reply:
Yes, `OpResult`/`OpOperand` are married to the SSA graph, but so does this. By specifying the idx for the src and dst, you are replicating the functionality of them (literally -- look at their implementations). I'm also not wild about having duplicate def-use specifications -- what if they get out of sync?

I suggest a public API which allows clients to inform the scheduler about the non-SSA dependences of side-effectful operations. (Which could be represented by (src,dst) operation pairs.) It would then be an implementation detail (possibly with some private helper functions) which extract and add the SSA dependencies. I think that would be a more appropriate place for a unified dependence model.

Reply:
Thanks for clarifying! I like that idea and will rework the API.

Reply:
Done (I think), see the new commits and my comment below (in the main conversation stream).